qualtran/bloqs/block_encoding.py

```diff
@@ -182,13 +186,46 @@ def pretty_name(self) -> str:
         return 'Prep'
 
 
+class BlockEncoding(Bloq):
+    r"""Abstract interface for an arbitrary block encoding.
+
+    Users must provide methods for defining the ancilla, junk, and system registers,
+    and a method for defining the signal state $|G\rangle$. See module level
+    docstrings for more discussion.
```

Comment:
Can we add more docstrings that explain the terms like "a", "signa" etc.? It woul probably mean copying some of the definitions from the linked reference but I think its worth doing

Reply:
I think this is a better reference for us to base the block encoding base class upon - https://arxiv.org/pdf/1804.01973

The block encoding base class should expose - (a) the state (or the registers and a bloq to prepare the state) `|G>` (b) the normalization constant `alpha` (in LCU, this will be the sum of coefficients) and (c) An error parameter `epsilon` -- these are parameters (usually denoted as `(a, alpha, epsilon)` are often used to define a block encoding). 

The other interesting case of assymetric block encodings where we have a different sized input (a) / output (b) registers can be described by our base class by just returning (one or more) registers which are not all necessarily THRU registers (but it'd be nice to verify that this is the case once we get to it). 

I don't think we need "junk" registers etc. ?  The registers to reflect upon would just correspond to the registers corresponding to `|G>` state and for block encodings which are not directly an LCU (eg SF / DF) the |G> state would be |0> and the registers would correspond to the "reflection registers" I believe. 

Does this make sense?

Reply:
epsilon is the precision for state preparation in this case?

Reply:
The block encoding still needs a signature though, should it just be ancilla and system?

Reply:
or selection and system.

Reply:
@tanujkhattar I took another stab at it